Add pod labels and annotation

ptitzler opened the pull request:
This PR
 - changes the `NotebookOp` constructor signature, adding
   - `pipeline_name` (required, no default)
   - `pipeline_version` (optional, default: `<empty string>`)
   - `experiment_name`  (required, no default) 
 - adds pod label(s)
   -  `elyra/node-type` => always `notebook-script` until Pipelines 2.0
   - `elyra/pipeline-name`
   - `elyra/pipeline-version`
   - `elyra/experiment-name`
   - `elyra/node-name`
 - adds pod annotation(s)
   - `elyra/node-file-name`
  - updates existing tests to accomodate constructor signature change
  - adds static class method `_normalize_label_value` 
 - adds tests for above method
 - partially solves https://github.com/elyra-ai/elyra/issues/1277 (another PR is required to handle pipeline export to Python DSL)
 - temporarily breaks elyra-ai/elyra due to `NotebookOp` signature change
 
With this enhancement administrators can
- identify which pods are associated with Elyra pipelines (in general)
- identify the Python script | notebook that's associated with the pod

### Labels metadata:
- `elyra-node-type`: static string identifying the node type as `notebook-script`, which covers file-based Elyra container ops for notebooks and Python scripts
- `elyra-pipeline-name`: the pipeline's experiment name, e.g. `my-labeled-pipeline3-0208165033` which is not the same as the pipeline name
- `elyra-node-name`: the pipeline node name (see also https://github.com/elyra-ai/elyra/issues/971)

### Annotations metadata:
- `elyra-node-file-name`: the notebook or Python script name, e.g. `notebook1.ipynb`

### How to test:

- Build and run
  ```
  $ make clean install 
  $ export ELYRA_GITHUB_BRANCH=i1277
  $ export ELYRA_GITHUB_ORG=ptitzler
  $ jupyter lab
  ```
#### Testing pipeline execution

- Create a pipeline and add any number of nodes
- Run the pipeline on KFP
- Log on to the KFP cluster 
- Run the following commands, replacing `[NAMESPACE]` with your namespace name
   ```
   # should return one entry for every node in every pipeline
   $ kubectl get pods -l=elyra-node-type --namespace=[NAMESPACE]
   # # should return one entry for every node in experiment EXPERIMENT_NAME
   $ kubectl get pods --selector=elyra-pipeline-name=[EXPERIMENT_NAME] 

   # should return one entry for every file-based node in the pipeline (results will differ once
   # Elyra supports Kubeflow components, which would be of a different type
   $ kubectl get pods --selector=elyra-node-type=notebook-script --namespace=[NAMESPACE]

   # for each [POD_NAME]  entry returned by the queries above run
   $ kubectl describe pods [POD_NAME] --namespace=[NAMESPACE]
   ```
 
#### Testing pipeline export to YAML
- Create a pipeline and add any number of nodes
- Export the pipeline to YAML
- Inspect the generated YAML file and verify that the following entries are present for each node (`[FILENAME]` and `[NODE_NAME]` are actual values from your pipeline)
   ```
   metadata:
        annotations: {elyra-node-file-name: [FILENAME]}
        labels: {elyra-node-type: notebook-script, elyra-node-name: [NODE_NAME], elyra-pipeline-name: [EXPERIMENT_NAME]}
   ```


Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

ptitzler commented on kfp_notebook/pipeline/_notebook_op.py:
```diff
@@ -228,6 +228,18 @@ def __init__(self,
+        # Pipeline name
+        # Should we use kwargs.get('pipeline_name')?
+        self.add_pod_label('elyra-pipeline-name',
+                           self.cos_directory)
```
1. The pipeline _experiment_ name is passed  as parameter `cos_directory`. If for whatever reason the value no longer represents the _experiment_ name, the attached metadata is invalid. For example, let's say in the future the `cos_directory` value is a UUID then this breaks.
2. I would suggest to introduce a required parameter `experiment_name` and use that one instead in this place. This does require changes in the `kfp_processor`.

ptitzler replied:
See https://github.com/elyra-ai/elyra/blob/master/elyra/pipeline/processor_kfp.py#L53 why `cos_directory` does not represent the pipeline name that the user entered.

kevin-bates replied:
I agree.  We need to stop overloading items here and this should be revisited.

Should the experiment name be the timestamp-suffixed value of pipeline name or just pipeline name?  I kinda think it should represent an _experiment_ or _run_ of a pipeline and therefore the former and then add a label for `elyra-experiment` so that all pods relative to a given run could be queried, as well as all pods for a given pipeline (that spans experiments).

ptitzler replied:
Are you saying we should expose both - the pipeline name and the experiment name? While they are currently identical, that might (and hopefully will) change in the future when we take advantage of pipeline versioning.

kevin-bates replied:
I view the pipeline name to be what the user entered and the experiment name to be the timestamp-suffixed value that is essentially _system-owned_ and changes on each submission/run - both of which are meaningful from a labeling perspective.

> when we take advantage of pipeline versioning

I would hope we'd introduce a new property to reflect a pipeline's version.

ptitzler replied:
Should be resolved. Introduced `pipeline_version` as an optional input parameter, which defaults to the empty string

ptitzler commented on kfp_notebook/pipeline/_notebook_op.py:
```diff
@@ -228,6 +228,18 @@ def __init__(self,
+        self.add_pod_label('elyra-pipeline-name',
+                           self.cos_directory)
+        # Pipeline node name
+        self.add_pod_label('elyra-node-name', kwargs.get('name'))
```
I forgot that we need to sanitize user-provided input to make sure it adheres to the _value_ constraints Kubernetes imposes. Do we already have a general purpose utility "class" that I could augment with two methods that convert a string to a valid kubernetes label/annotation value?
The keys I'm using are static and therefore don't require any processing.

kevin-bates replied:
Good point.  I doubt anything like that exists.  However, [annotation values](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set) do not adhere to the same constraints as [label values](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).

> If the prefix is omitted, the annotation Key is presumed to be private to the user. **Automated system components** (e.g. kube-scheduler, kube-controller-manager, kube-apiserver, kubectl, or **other third-party automation) which add annotations to end-user objects must specify a prefix**.

Seems like we should treat `elyra` as the prefix portion for the label and annotation keys: e.g., `elyra/node-node`, `elyra/pipeline-name`

ptitzler replied:
Done
 - Added a static class method that validates the label value. No validation performed for annotation values because I couldn't find a formal definition of what constraints are enforced.
 - Used `elyra` prefix as suggested, although it'd be better if there was an `elyra.io` (or similar) DNS entry that the elyra project owns, that ties the prefix to something people can look up

ptitzler commented:
**Note this is a breaking change** and should only be merged after it's companion PR in elyra-ai/elyra has been reviewed and is ready for merge.

kevin-bates commented on kfp_notebook/pipeline/_notebook_op.py:
```diff
@@ -236,3 +266,40 @@ def _artifact_list_to_str(self, pipeline_array):
+        # must begin with [0-9a-zA-Z]
+        valid_chars = string.ascii_letters + string.digits
+        if value[0] not in valid_chars:
+            value = 'a' + value[1:]
```
Do we want to _replace_ the first character, rather than just prepend to it?  Since its arguably the most significant character, I'm thinking it should be preserved - albeit illegal as a label value.

Of course, the truncation would need to happen after the first-character check.

ptitzler replied:
Let's use example `#1234567890...YZ`
 - prepend `a` (`a#1234567890...YZ`) and
 - truncate if new total length exceeds the max 63 `a#1234567890...Y` OR
 - always truncate to retain original length `a#1234567890...Y`
WDYT?

ptitzler replied:
I guess in the spirit of your initial comment it would make most sense to preserve the original string as much of possible and only truncate if the length constraint is violated

kevin-bates replied:
yes.  Only truncate to 63 if the _current_ length (i.e., after possible prefixing) is > 63.

kevin-bates commented on kfp_notebook/pipeline/_notebook_op.py:
```diff
@@ -236,3 +266,40 @@ def _artifact_list_to_str(self, pipeline_array):
+            value = 'a' + value[1:]
+
+        # must end with [0-9a-zA-Z]
+        if value[len(value) - 1] not in valid_chars:
+            value = value[:len(value) - 1] + 'a'
```
A more "pythonic" way might be...
```suggestion
        if value[-1] not in valid_chars:
            value = value[:-1] + 'a'
```

kevin-bates reviewed (commented):
This is looking good Patrick.  Just had a question and comment.

The label value test is very nice!

ptitzler commented:
I've updated the processing logic based on the review feedback and adjusted the test cases accordingly.

kevin-bates reviewed (approved):
This looks good.  I like `valid_char` and `valid_middle_char` over the three replacement forms since it's not always a replacement, but that's getting a little nit-picky.

This will be useful and fairly easy to extend to allow custom labels/annotations down the road.  Thanks.

ptitzler commented:
> I like valid_char and valid_middle_char over the three replacement forms since it's not always a replacement, but that's getting a little nit-picky.

I changed the variable names and added documentation.

ptitzler commented:
Sibling PR https://github.com/elyra-ai/elyra/pull/1284 is now under review

akchinSTC reviewed (approved):
LGTM, run on kfp w/ argo and tekton
